### PR TITLE
Integrate libspector

### PR DESCRIPTION
**Pull request by j0ni**
This introduces libspector, and supports two use cases:

1. Configure match text in the `agent.conf` file, and track processes which match.
2. Always run a periodic scan of all running processes.

There's a companion coalmine PR appcanary/coalmine#196 which adds datamodel for the second use case, as well as an endpoint. Not sure there's much point in keeping the first part of the implementation here, as configuring it will simply duplicate a subset of the data generated by the second use case.

### CODE REVIEW

File: agent/conf.go [hunk 1]
```diff
@@ -19,6 +19,7 @@ type Conf struct {
 
 type FileConf struct {
 	Path    string `toml:"path"`
+	Command string `toml:"command"`
```
Comment (j0ni) [inline agent/conf.go]: NB this is a UI change, in that `command` means what `process` used to mean. I think the semantics are clearer with this name, but maybe I should leave `process` with the old semantics and pick a new name for the process watching config. Or even maybe we don't want a string match based implementation at all...

File: agent/processwatcher.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+func NewAllProcessWatcher(callback ChangeHandler) Watcher {
+	return NewProcessWatcher("", callback)
+}
+
+func (wt *processWatcher) Start() {
+	wt.Lock()
+	wt.keepPolling = true
+	wt.Unlock()
+	go wt.listen()
```
Comment (j0ni) [inline agent/processwatcher.go]: Deferring the unlock until after the go routine is dispatched produced a deadlock most of the time. I'm really not sure about the value of all this coordination.

File: test/data/spector.conf [hunk 3]
```diff
@@ -0,0 +1,8 @@
+server_name = "spectortest"
+
+api_key = "XtG6yAdbsSJfMwgx9SwafhZ8"
```
Comment (j0ni) [inline test/data/spector.conf]: I changed this :blush:

File: agent/conf.go (hunk 1 above)
Comment (phillmv) [inline agent/conf.go]: This is fine, we just have to support all existing configuration forever.

File: agent/processwatcher.go [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+func (w watchedState) String() string {
+	buffer := bytes.NewBufferString("Watched Processes:\n\n")
+	for _, proc := range w {
+		buffer.WriteString(fmt.Sprintf("%v\n", proc.String()))
+	}
+	return buffer.String()
+}
+
+type watchedProcess struct {
+	ProcessStarted time.Time
```
Comment (phillmv) [inline agent/processwatcher.go]: minor style: ProcessStartedAt

File: agent/processwatcher.go [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+	if w.Outdated {
+		buffer.WriteString(", is running outdated lib(s)")
+	}
+	for _, lib := range w.Libraries {
+		buffer.WriteString(fmt.Sprintf("\n%v", lib.String()))
+	}
+	return buffer.String()
+}
+
+type library struct {
```
Comment (phillmv) [inline agent/processwatcher.go]: it'd be amaze if we had the same name across projects

File: agent/processwatcher.go [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+		if err != nil {
+			log.Warningf("error retrieving modification date for lib %s, %v", path, err)
+			continue
+		}
+
+		var fullPackage string
+		if pkg, err := lib.SpectorLib.Package(); err != nil {
+			log.Warningf("error retrieving package name for lib %s, %v", path, err)
+		} else {
+			fullPackage = fmt.Sprintf("%s-%s", pkg.Name(), pkg.Version())
```
Comment (phillmv) [inline agent/processwatcher.go]: this may or may or not matter but make sit a bit simpler on receiving end if we send "package_name" and "package_version"

File: agent/processwatcher.go [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+
+		state[proc.PID()] = wp
+	}
+
+	return &state
+}
+
+func (wt *processWatcher) scan() {
+	wt.Lock()
+	wt.state = wt.acquireState()
```
Comment (phillmv) [inline agent/processwatcher.go]: potentially check CRC, thereby fully bringing processWatcher into Watcher family.

File: agent/textwatcher.go [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+	CmdArgs      []string
+	contents     func() ([]byte, error)
+	pollSleep    time.Duration
+}
+
+// File watchers track changes in the contents of a file
+func NewFileWatcher(path string, callback ChangeHandler) Watcher {
+	var kind string
+	filename := filepath.Base(path)
+	switch filename {
```
Comment (phillmv) [inline agent/textwatcher.go]: add a TODO about getting rid of this once we fix it in coalmine.

Similarly, add a todo about sending along the platform and release with every update/onchange.

File: agent/textwatcher.go [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+// Process watchers track changes in the output of a command
+func NewCommandOutputWatcher(process string, callback ChangeHandler) Watcher {
+	splat := strings.Split(process, " ")
+	name := splat[0]
+	args := splat[1:]
+
+	watcher := &textWatcher{
+		path:      process,
+		OnChange:  callback,
+		kind:      "centos",
```
Comment (phillmv) [inline agent/textwatcher.go]: lol

File: test/data/test.conf [hunk 10]
```diff
@@ -12,4 +12,4 @@ path = "/var/lib/dpkg/available"
 path = "/path/to/Gemfile.lock"
 
 [[files]]
-process = "fakecmdhere"
+command = "fakecmdhere"
```
Comment (phillmv) [inline test/data/test.conf]: No. Bad!

**Comment by phillmv** [comment]
1. Existing configurations files must not break!
2. Rename to libraries ProcessLibrary to match our end.
3. That minor todo.
4. Split up the package-version into two attributes
5. CRC check maybe?

Otherwise looks great.

File: agent/textwatcher.go (hunk 8 above)
Comment (j0ni) [inline agent/textwatcher.go]: Not sure what the first part of this refers to. The second part we have been discussing - I added and then removed these items from the payload. I can easily add them back if we want to update them with every newly received process watcher payload.

**Comment by j0ni** [comment]
I implemented the CRC check in the crudest way, and because the `MarshalJSON` code makes calls to `libspector` to resolve package and package version, this results in duplicate calls occurring per run.

To continue using the CRC check, I think I should either pre-load all that data, or instead of capturing the data structure, capture the JSON blob of bytes.

File: agent/processwatcher.go [hunk 11]
```diff
@@ -0,0 +1,434 @@
...
+	for i, library := range pm.libraries {
+		if library.Path() == path {
+			return i
+		}
+	}
+
+	return -1
+}
+
+func (pm *processMap) maybeAddLibrary(lib libspector.Library) int {
```
Comment (phillmv) [inline agent/processwatcher.go]: "maybe" here denotes that the func does a set operation - won't add it if it already exists.

I guess I expect that to mean it could sometimes work, sometimes not work :P. Perhaps just "addLibrary"?

File: agent/processwatcher.go [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+
+	if index < 0 {
+		pm.libraries = append(pm.libraries, lib)
+		index = pm.findLibraryIndex(path)
+	}
+
+	return index
+}
+
+func (wt *processWatcher) acquireState() *processMap {
```
Comment (phillmv) [inline agent/processwatcher.go]: s/wt/pw/, and move below the initializer?

File: agent/processwatcher.go [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+	return wt.match
+}
+
+func (wt *processWatcher) KeepPolling() bool {
+	wt.Lock()
+	defer wt.Unlock()
+	return wt.keepPolling
+}
+
+func (wt *processWatcher) setStateAttribute() {
```
Comment (phillmv) [inline agent/processwatcher.go]: s/wt/pw/?

File: agent/processwatcher.go [hunk 14]
```diff
@@ -0,0 +1,434 @@
...
+		UpdatedAt: time.Now(),
+		pollSleep: env.PollSleep,
+	}
+
+	// Don't scan from here, we just end up with two running at once
+	return watcher
+}
+
+func NewAllProcessWatcher(callback ChangeHandler) Watcher {
+	return NewProcessWatcher("", callback)
```
Comment (phillmv) [inline agent/processwatcher.go]: Let's make this `"*"` instead of `""` - seems easier to not do accidentally :).

File: agent/processwatcher.go [hunk 15]
```diff
@@ -0,0 +1,434 @@
...
+			}
+			wp.Libraries[i] = lib
+		}
+
+		pm.processes = append(pm.processes, wp)
+	}
+
+	// to make the crc check more meaningful
+	sort.Sort(pm.processes)
+	sort.Sort(pm.libraries)
```
Comment (phillmv) [inline agent/processwatcher.go]: I see you're comparing things on `lib.Path()`.

When we add something to pm.libraries, we iterate thru the list and return with an index when `lib.Path()` matches. Won't this scramble the index stored in `processLibrary`?

File: agent/textwatcher.go (hunk 8 above)
Comment (phillmv) [inline agent/textwatcher.go]: first part: the kind attribute we receive in the payload is kind of useless. If the agent just looks at the file name in order to set the kind, we can do that just fine in the server also.

the second part: we can do it as a separate pr.

File: agent/textwatcher.go (hunk 9 above)
Comment (phillmv) [inline agent/textwatcher.go]: i.e. see above re: kind being useless. It'd make more sense to set kind to be the _kind of watcher_.

This will matter once we support a non-ruby, non-linux.

File: agent/processwatcher.go [hunk 16]
```diff
@@ -0,0 +1,434 @@
...
+	go wt.listen()
+}
+
+func (wt *processWatcher) Stop() {
+	wt.Lock()
+	wt.keepPolling = false
+	wt.Unlock()
+}
+
+func (wt *processWatcher) Match() string {
```
Comment (phillmv) [inline agent/processwatcher.go]: I feel this could be more descriptive somehow, esp with how it's reflected in the configuration.

```Go
	Command string `toml:"process"`
	Process string `toml:"match"`
```

This is already pretty "lol", since we have to backwards compat it. But how do you feel about (just throwing it out there), `inspect_process`?

Comment (phillmv) [inline agent/processwatcher.go]: we could also add "command" to be the new canonical Command string and change docs, while silently supporting/warning on old use.

File: agent/processwatcher.go [hunk 17]
```diff
@@ -0,0 +1,434 @@
...
+// 		libraries: &systemLibraries{
+// 			someLib1,
+// 			someLib2,
+// 			// ...
+// 		},
+// 	},
+// }
+
+// "map" in the colloquial sense
+type processMap struct {
```
Comment (phillmv) [inline agent/processwatcher.go]: what if instead of 'processMap' we called this 'systemState'? I also spent a bunch of time parsing out the diff b/w processMap and processWatcher.

File: agent/processwatcher.go (hunk 16 above)
Comment (j0ni) [inline agent/processwatcher.go]: `inspect_process` seems too verby to me, but `process_to_inspect` is cumbersome and I can't think of another. Unless, I guess we could switch to `inspect_process` and `exec_process`. But we have the same legacy issue there.

File: agent/processwatcher.go (hunk 15 above)
Comment (j0ni) [inline agent/processwatcher.go]: Good catch, a proper bug. I think I resolved it with a more thoughtful sort process.

File: agent/processwatcher.go [hunk 18]
```diff
@@ -0,0 +1,434 @@
...
+
+		command, err := proc.Command()
+		if err != nil {
+			log.Infof("Can't read command line for PID %d: %v", proc.PID(), err)
+			// fall through, we can live without this (?)
+		}
+
+		spectorLibs, err := proc.Libraries()
+		if err != nil {
+			if os.Getuid() != 0 && os.Geteuid() != 0 {
```
Comment (j0ni) [inline agent/processwatcher.go]: That last case is the one I was considering, but you're right.

Would you rather just panic on startup?

OTOH, right now this is the difference between a test run that succeeds, and one which crashes hard.

File: agent/processwatcher.go (hunk 12 above)
Comment (j0ni) [inline agent/processwatcher.go]: Not sure I understand what you mean by initializer - are you talking about the 2 constructor functions?

**Comment by phillmv** [comment]
LOOKS GOOD TO PHILLIP
